source/MRViewer/MRSubfeaturePickerWidget.h

```diff
@@ -0,0 +1,218 @@
+    // Whether ancillary features should be processed.
+    bool allowAncillaryObjects = false;
+
+    MRVIEWER_API static float defaultInfiniteExtent();
+    // How large we should make "infinite" features.
+    std::function<float()> getInfiniteExtent = defaultInfiniteExtent;
+
+    // Temporarily override feature sizes to those values for picking:
+    float pickerPointSize = 16;
+    float pickerLineWidth = 10;
+
+    struct SubfeatureData
+    {
+        std::string name;
+        Features::Primitives::Variant subfeature;
+        std::shared_ptr<FeatureObject> object;
+    };
+
+    struct ObjectData
+    {
+        Features::Primitives::Variant feature;
+        std::vector<SubfeatureData> subfeatures;
+    };
+
+    using ObjectMap = std::map<std::shared_ptr<Object>, std::shared_ptr<ObjectData>>;
+    // Returns all tracked feature objects.
+    [[nodiscard]] MRVIEWER_API const ObjectMap& getObjects() const;
+
+    // Update the temporary objects representing subfeatures.
+    MRVIEWER_API void updateTemporaryObjects( ObjectPred objectPredicate = nullptr );
+
+    // Remove all temporary objects. The next `updateTemporaryObjects()` will add them back.
+    // Do this before disabling your plugin.
+    MRVIEWER_API void removeTemporaryObjects();
+
+    // Runs the picker for all visible pickable objects, including subfeatures.
+    // Uses `temporarilyAdjustSubfeaturesForPicking()` under the hood.
+    // NOTE! If you're going to call this from `mouseMove`, don't forget `if ( ImGui::GetIO().WantCaptureMouse ) /*pick nothing*/;` in `drawDialog_()`.
+    [[nodiscard]] MRVIEWER_API ObjAndPick pickObject( std::function<bool( const Object& )> pred = nullptr );
+
+    // Temporarily unhides the subfeature objects that are normally hidden, and also temporarily sets their radius and line width to improve picking.
+    // Then calls `func()` and reverts everything back..
+    // You normally don't need this funcion, prefer `pickObject()` if possible.
+    MRVIEWER_API void temporarilyAdjustSubfeaturesForPicking( std::function<void()> func );
+
+    struct FindResult
+    {
+        // Both bools can only be null at the same time.
+        const ObjectMap::value_type* feature = nullptr;
+        const SubfeatureData* subfeature = nullptr;
+
+        [[nodiscard]] explicit operator bool() const { return feature && subfeature; }
+    };
+
+    // If `subfeature` is one of our subfeatures, returns the information about it, including the parent feature. Otherwise returns null.
+    [[nodiscard]] MRVIEWER_API FindResult findSubfeature( const Object& subfeature );
+
+
+private:
+    // Using `std::shared_ptr` for the key here to be able to attach signal lifetimes to this.
+    ObjectMap objects_;
+};
+
+// A helper class to pick into subfeatures.
+// Highlights hovered objects, and/or draws GUI to further descend into subfeatures.
+// Workflow:
+// 1. Add this as a data member to your plugin. Also add `ObjectHighlighter`.
```

Comment:
it has internal ObjectHighliter, does it need another one in plugin?

Reply:
Made singleton.